ESQL: Move more test type error testing

nik9000 opened the pull request:
This reduces the number of test cases in ESQL a little more ala #119678. It migrates a few random tests and all of the multivalue functions:
```
92775 -> 43760
 3m45 -> 4m04
```

This adds a few more error test cases that were missing to make sure it all lines up well. And it fixes a few error messages in a few functions. That's *likely* where the extra time goes.

nik9000 committed:
ESQL: Move more test type error testing

This reduces the number of test cases in ESQL a little more ala #119678.
It migrates a few random tests and all of the multivalue functions:
```
92775 -> 43760
 3m45 -> 4m04
```

This adds a few more error test cases that were missing to make sure it all
lines up well. And it fixes a few error messages in a few functions. That's
*likely* where the extra time goes.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 committed:
Merge branch 'main' into move_error_cases3

not-napoleon commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MatchErrorTests.java:
```diff
@@ -0,0 +1,75 @@
+            return typeErrorMessage(true, validPerPosition, signature, positionalErrorMessageSupplier);
+        } catch (IllegalStateException e) {
+            // This means all the positional args were okay, so the expected error is for nulls or from the combination
+            return EsqlBinaryComparison.formatIncompatibleTypesMessage(signature.get(0), signature.get(1), sourceForSignature(signature));
```
Out of scope here, but maybe we should move `formatIncompatibleTypesMessage` if we're going to use it for more than just the comparison operators.

not-napoleon commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvCountErrorTests.java:
```diff
@@ -0,0 +1,53 @@
+             * In general MvCount should support all signatures. While building a
+             * new type you may we to temporarily remove this.
+             */
+            throw new UnsupportedOperationException("all signatures should be supported");
```
I like this pattern for validating that all types work.  Let's add a note in the javadoc for `ErrorsForCasesWithoutExamplesTestCase` that we want to follow this pattern for "any type" functions.  If it comes up enough, might be worth refactoring it too.

not-napoleon commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSortTests.java:
```diff
@@ -36,12 +36,28 @@ public MvSortTests(@Name("TestCase") Supplier<TestCaseSupplier.TestCase> testCas
+            nulls(suppliers, orderType);
+        }
+
+        List<TestCaseSupplier> extra = new ArrayList<>();
```
This could use a comment about what exactly it's adding cases for.

not-napoleon reviewed (approved):
LGTM.  I left a few notes about possible follow-up work, but this is a big improvement IMHO.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSort.java:
```diff
@@ -126,7 +127,10 @@ protected TypeResolution resolveType() {
+        TypeResolution resolution = isType(field, dt -> switch (dt) {
+            case UNSIGNED_LONG, GEO_POINT, GEO_SHAPE, CARTESIAN_POINT, CARTESIAN_SHAPE -> false;
+            default -> true;
+        }, sourceText(), FIRST, "not unsigned long or geo");
```
I think that's out of scope, yeah, but it does feel useful.

ivancea reviewed (approved):
LGTM! Some minor concerns, _ignore them at will_

nik9000 committed:
Examples

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSort.java:
```diff
@@ -136,7 +140,7 @@ protected TypeResolution resolveType() {
         }
 
-        return isString(order, sourceText(), SECOND);
+        return isNotNull(order, sourceText(), SECOND).and(isString(order, sourceText(), SECOND));
```
Got it. I'll see if I can incorportate that bit.

nik9000 committed:
ESQL: Heap attack tests

Adds tests for requests that would fill up the heap and crash
elasticsearch but for our circuit breakers. In an ideal world we'd
stream these results back to and this wouldn't crash anything. But we
don't have that at the moment.

nik9000 committed:
Merge branch 'main' into esql_lookup_heap_attack

nik9000 committed:
Merge branch 'main' into move_error_cases3

nik9000 committed:
Merge remote-tracking branch 'nik9000/esql_lookup_heap_attack' into move_error_cases3

nik9000 committed:
REvert more

nik9000 committed:
How'd this get in there?

nik9000 committed:
Revert harder

nik9000 committed:
Revert proper

nik9000 reviewed (commented):
I've reverted the troublesome changes - `MvPercentile` and `MvSort`. Those we'll grab in another PR.

nik9000 commented:
Backport: https://github.com/elastic/elasticsearch/pull/120324
